contrib/egrul/generate.py

```diff
@@ -0,0 +1,403 @@
+from collections import defaultdict
+from datetime import date, datetime, timedelta
+from pathlib import Path
+from typing import Optional, List, Iterable, Dict
+from zipfile import ZipFile
+
+import numpy as np
+import pandas
+from pyspark import Row, StorageLevel
+from pyspark.sql import SparkSession
+from pyspark.sql.connect.dataframe import DataFrame
+from pyspark.sql.functions import (
+    struct,
+    udf,
+    col,
+    coalesce,
+    explode_outer,
+    explode,
+)
+
+from egrul_xml import parse_xml
+from schema import company_record_schema
+from zavod import Context
+from zavod import Dataset
+
+LOCAL_BUCKET_PATH = "/Users/leon/internal-data/"
+INTERNAL_DATA_ARCHIVE_PREFIX = "ru_egrul/egrul.itsoft.ru/EGRUL_406/"
+INTERNAL_DATA_CACHE_PREFIX = "ru_egrul/cache"
+INTERNAL_DATA_PROCESSESED_PREFIX = "ru_egrul/processed/"
+
+
+def day_before(d: date) -> date:
+    return d - timedelta(days=1)
+
+
+def update_company_from_new_company(context: Context, old: Row, new: Row) -> Row:
+    """Enriches a company with information from a previous version."""
+    # Doing this in Python is too expensive, we use a Spark join
+    assert old is not None and new is not None, "Both old and new must be present"
+
+    result = new
+    data_date = new.legal_entity.seen_date
+
+    # assert all(
+    #     [o.start_date is not None for o in result.ownerships]
+    # ), "Ownership does not have a start_date"
+
+    expired_ownerships = []
+    new_ownership_ids = set([o.id for o in result.ownerships])
+    for o in old.ownerships:
+        # Ownership.id is built from (owner, asset, shares_count, role), so any change will build a new Ownership
+        if o.id not in new_ownership_ids:
+            o_dict = o.asDict()
+            o_dict["end_date"] = day_before(data_date)
+            expired_ownerships.append(Row(**o_dict))
+
+    expired_directorships = []
+    new_directorship_ids = set([o.id for o in result.directorships])
+    for d in old.directorships:
+        # Directorship.id is built from (company, director, role)
+        if d.id not in new_directorship_ids:
+            # We try to find a new directorship with the same role, and if it has a start_date,
+            # we use that as a better end_date for the previous directorship than just the day we saw it.
+            new_directorship_same_role = [
+                new_d for new_d in result.directorships if new_d.role == d.id
+            ]
+            d_dict = d.asDict()
+            if (
+                len(new_directorship_same_role) > 0
+                and new_directorship_same_role[0].start_date is not None
+            ):
```

Comment:
same as for ownership - does this need a same-owner check?

Reply:
See above, the trick is that we're checking on ID. But maybe I'm also not thinking about it right?